Review for changes to specification for Unified NumberFormat

sffc opened the pull request.

sffc committed:
Adding more background and sections to README file

sffc committed:
Adding section to spec for unit resolution.

sffc committed:
Adding definitions.

sffc committed:
Refactoring pattern resolution into its own function that is dependent on the number being formatted.

sffc committed:
Adding measure unit fields and adding abstraction for plural form.

sffc committed:
Adding plumbing for notation setting; the setting itself is not yet used.

sffc committed:
Removing redundant infinity ILD String line.

sffc committed:
Adding some more information about measure units and notation to README

sffc committed:
Adding even more to the README and adding 'narrowSymbol' value to the currencyDisplay enum.

littledan commented on README.md:
```diff
@@ -1,19 +1,127 @@
+## Background / Motivation
 
-There are many requests for adding number-formatting-related features to ECMA 402.  A few of them include:
+There are many requests for adding number-formatting-related features to ECMA 402. These include:
```
For background, write "ECMA-402 (the JavaScript Intl standard library)" to make this more self-contained.

sffc replied:
Done.

littledan commented on README.md:
```diff
@@ -1,19 +1,127 @@
+- [Add scientific notation](https://github.com/tc39/ecma402/issues/164) -- Section III
+- [Add option to force sign](https://github.com/tc39/ecma402/issues/163) -- Section IV
+- [Add compact decimal notation](https://github.com/tc39/ecma402/issues/37) -- Section III
+- [Add measure unit formatting](https://github.com/tc39/ecma402/issues/32) -- Section II
```
Minor aesthetic suggestion: Sort these by section number

sffc replied:
Done.

littledan commented on README.md:
```diff
@@ -1,19 +1,127 @@
+
+Rather than complicate `Intl` with more constructors with heavilly overlapping functionality, this proposal is to restructure the spec of `Intl.NumberFormat` to make it more easilly support additional features in a "unified" way.
+
+Additional background: [Unified API for number formatting](https://github.com/tc39/ecma402/issues/215)
```
Minor: Re-label "`[prior discussion](url)`"?

sffc replied:
Done.

littledan commented on README.md:
```diff
@@ -1,19 +1,127 @@
+
+Additional background: [Unified API for number formatting](https://github.com/tc39/ecma402/issues/215)
+
+## I. Spec Cleanup
```
I like how the later sections include code samples demonstrating their use. Do you think you could add a code sample here using `currencyDisplay: "narrowSymbol"`?

I don't think it's necessary to include the unobservable spec refactorings in this document. If you want to be clear about the changes, I'd recommend making this a PR of 5 patches, each of which has a detailed commit message including things about the editorial/refactoring changes. Right now, it doesn't really look like the patches correspond to these 5 bullet points.

sffc replied:
Added sample code for narrowSymbol.

Which 5 bullet points?

The commits on this PR are not intended to correspond to sections of the README file proposal.  Right now the changes for the four sections are somewhat intertwined, so separating them would require some thought.  However, if you think the committee would be more amenable if the changes were separated, I can work on doing that.

littledan replied:
Sorry, I just meant, "if". In my opinion, the most important thing is the observable changes, and you can just omit the description of the spec refactoring. (By bullets, I just meant the `## I.` headings.)

littledan commented on README.md:
```diff
@@ -1,19 +1,127 @@
+The syntax was discussed in #3.
+
+- `style` receives the string value "unit"
+- `unit` receives a string measure unit identifier, defined in [UTS #35](http://unicode.org/reports/tr35/tr35-general.html#Unit_Elements).  See also the [full list of unit identifiers](https://unicode.org/repos/cldr/tags/latest/common/validity/unit.xml).
```
@waldemarhorwat has previously objected to a previous unit list as it wasn't arbitrary generative. It's good that you have the cross-reference to CLDR here; this rationale (that this is the shape of the data we have) may be good to highlight in the presentation.

sffc replied:
Added a note about this.  What do you mean by "arbitrary generative"?

littledan replied:
I mean, not allowing unrestricted numerator and denominator units which are each products of other units.

sffc replied:
Thanks for the clarification.  I reworded my discussion.  Since proper support for compound units requires more design work and popular compound units like meters per second are already built in to CLDR directly, I do not plan to add them to this proposal.

littledan replied:
That sounds good to me. If you can include this explanation in your presentation, it will be helpful.

littledan commented on README.md:
```diff
@@ -1,19 +1,127 @@
+The syntax was discussed in #6.
+
+- `signDisplay` takes either "auto", "always", "never", or "except-zero"
+- `currencySignDisplay` takes either "standard" or "accounting"
```
Can you explain here what these various options values mean, and what the default behavior is when they are missing? I don't think everyone reading this will know that "accounting" stands for using the parentheses, and "standard" stands for a - sign. Examples in non-en-US locales would also be nice.

sffc replied:
I added a table for the signDisplay options.

I fleshed out the paragraph about accounting format a bit more and changed it to a non-en-US locale.

littledan commented on README.md:
```diff
@@ -1,19 +1,127 @@
+
+Rounding-related settings (min/max integer/fraction digits) are applied after the number is scaled according to the chosen notation.
+
+Notation styles are allowed to be combined with other options.  (ICU supports this.)
```
Maybe best to omit comments like "ICU supports this"; the feature makes sense even for non-ICU-based implementations, and on the other hand you're not providing general mapping guidance in the other parts of this document.

sffc replied:
Done.

littledan reviewed (commented):
Here's some suggestions on the explainer; I'll come back and review the spec text a bit later.

sffc committed:
Responding to @littledan review feedback part 1

littledan commented on numberformat_proposed.html:
```diff
@@ -463,6 +482,84 @@ <h1>UnwrapNumberFormat( _nf_ )</h1>
+        1. If _style_ is `"unit"` or _style_ is `"percent"`, then
+          1. If _unit_ is *undefined*, throw a *TypeError* exception.
+          1. If the result of IsWellFormedUnitIdentifier(_unit_) is *false*, throw a *RangeError* exception.
+          1. Set _intlObj_.[[UnitType]] to `"measure-unit"`.
```
Editorial comment: It seems like the phrase "measure-unit" never leaks out into the public API (as we agreed on IIRC); maybe you could switch to "unit"/`[[Unit*]]` internally as well.

sffc replied:
I had considered that.  The reason I used "measure-unit" internally was because I also created an internal field "UnitType", which has values "currency", "measure-unit", or "none", which are each a specific class of units.  This internal vocabulary is influenced by ICU.

littledan commented on README.md:
````diff
@@ -1,19 +1,148 @@
+// ==> "$100" (rather than "US$100")
+```
+
+## II. Measure Units
````
"Units"?

sffc replied:
Done.

littledan commented on README.md:
```diff
@@ -1,19 +1,148 @@
+    currency: "EUR",
+    currencySignDisplay: "accounting"
+});
+// ==> (১০০.০০€)
```
Stupid question: Is this output really 55.00, and not 100.00?

sffc replied:
Not stupid at all.  Fixed.

littledan commented on numberformat_proposed.html:
```diff
@@ -463,6 +482,84 @@ <h1>UnwrapNumberFormat( _nf_ )</h1>
+        1. Let _dataLocaleData_ be _localeData_.[[&lt;_dataLocale_&gt;]].
+        1. Let _patterns_ be _dataLocaleData_.[[patterns]].
+        1. Assert: _patterns_ is a record (see <emu-xref href="#sec-intl.numberformat-internal-slots"></emu-xref>).
+        1. TODO: Is there a spec defining the format of _patterns_ that needs to be updated, too?
```
Do you mean the format of these nested records that include the patterns? I think this is what you'd want to update: https://tc39.github.io/ecma402/#sec-intl.numberformat-internal-slots

sffc replied:
Yes.  Thanks.

littledan commented on numberformat_proposed.html:
```diff
@@ -721,15 +842,22 @@ <h1>Properties of Intl.NumberFormat Instances</h1>
 
     <ul>
       <li>[[Locale]] is a String value with the language tag of the locale whose localization is used for formatting.</li>
+      <li>[[DataLocale]] is a String value with the language tag of the nearest locale for which the browser has data to perform the formatting operation.  (TODO: is this precisely correct??)</li>
```
In TC39, we talk about "implementations" or "JavaScript engines" rather than browsers (e.g., to take Node.js into account). Here's some alternate wording (which also adds a bit of context):

> \[\[DataLocale]] is a String value with the language tag of the nearest locale for which the implementation has data to perform the formatting operation. It will be a parent locale of \[\[Locale]].

sffc replied:
Great; thanks!

littledan commented on README.md:
```diff
@@ -1,19 +1,148 @@
+The syntax was discussed in #6.
+
+- `signDisplay`: "auto" (default), "always", "never", "except-zero"
+- `currencySignDisplay`: "standard" (default), "accounting"
```
Explaining currencySignDisplay here would be good (copy-pasting the explanation from the specification would work; many people who read the README won't read the spec).

sffc replied:
Added a pointer to the example above.

littledan commented on numberformat_proposed.html:
```diff
@@ -463,6 +482,84 @@ <h1>UnwrapNumberFormat( _nf_ )</h1>
+        1. Let _currency_ be ? GetOption(_options_, `"currency"`, `"string"`, *undefined*, *undefined*).
+        1. Let _currencyDisplay_ be ? GetOption(_options_, `"currencyDisplay"`, `"string"`, &laquo; `"code"`, `"symbol"`, `"narrowSymbol"`, `"name"` &raquo;, `"symbol"`).
+        1. Let _unit_ be ? GetOption(_options_, `"unit"`, `"string"`, *undefined*, *undefined*).
+        1. Let _unitDisplay_ be ? GetOption(_options_, `"currencyDisplay"`, `"string"`, &laquo; `"short"`, `"narrow"`, `"long"` &raquo;, `"short"`).
```
s/currencyDisplay/unitDisplay/?

sffc replied:
Thanks!  Done.

littledan commented on numberformat_proposed.html:
```diff
@@ -463,6 +482,84 @@ <h1>UnwrapNumberFormat( _nf_ )</h1>
+          1. Let _unit_ be `"concentr-percent"`.
+        1. If _style_ is `"unit"` or _style_ is `"percent"`, then
+          1. If _unit_ is *undefined*, throw a *TypeError* exception.
+          1. If the result of IsWellFormedUnitIdentifier(_unit_) is *false*, throw a *RangeError* exception.
```
Where is IsWellFormedUnitIdentifier defined?

sffc replied:
Good catch; I may have forgotten about this one.  It should be in same place as `IsWellFormedCurrencyCode`.  Looks like that is in section 6.3.1 of ECMA-402.  I guess I should add a section 6.5 about measurement unit codes with the new function `IsWellFormedUnitIdentifier` in section 6.5.1?

littledan replied:
Sounds good.

sffc replied:
Done in latest commit.

littledan replied:
Thanks for writing that out. Its semantics are different from what I expected!

littledan commented on numberformat_proposed.html:
```diff
@@ -721,15 +842,22 @@ <h1>Properties of Intl.NumberFormat Instances</h1>
-      <li>[[Currency]] is a String value with the currency code identifying the currency to be used if formatting with the `"currency"` style. It is only used when [[Style]] has the value `"currency"`.</li>
-      <li>[[CurrencyDisplay]] is one of the String values `"code"`, `"symbol"`, or `"name"`, specifying whether to display the currency as an ISO 4217 alphabetic currency code, a localized currency symbol, or a localized currency name if formatting with the `"currency"` style. It is only used when [[Style]] has the value `"currency"`.</li>
+      <li>[[Style]] is one of the String values `"decimal"`, `"currency"`, `"percent"`, or `"unit"`. This value corresponds closely with the [[UnitType]] value.</li>
+      <li>[[UnitType]] is one of the String values `"none"`, `"currency"`, or `"measure-unit"`. The value `"measure-unit"` is used for both [[Style]] `"percent"` and `"unit"`.</li>
```
Why is UnitType represented separately from Style? It seems like the only difference here is organizing "percent" as a kind of "measure-unit"--am I misunderstanding?

sffc replied:
Style is public; UnitType is internal.  UnitType maps more directly to data and ICU.

At some point I need to make the resolution from percent to measure-unit; it seems cleanest to do that at the API boundary and store the result in UnitType.

littledan replied:
OK, I can see how UnitType corresponds to what you have in your locale database, and Style corresponds to the options bag and resolvedOptions, but it still seems pretty redundant to me. I don't have an opinion whether it's stored in the object as Style or UnitType, but I'd prefer logic at the usage sites to translate from one to the other rather than redundantly storing the information.

littledan commented on numberformat_proposed.html:
```diff
@@ -402,7 +420,7 @@ <h1>ToRawPrecision( _x_, _minPrecision_, _maxPrecision_ )</h1>
           1. If the last character of _m_ is `"."`, then
             1. Remove the last character from _m_.
-        1. Return _m_.
+        1. Return the tuple [_m_, _xFinal_].
```
We don't tend to use "tuples" in the JavaScript specification. More typical would be to use Records, such as, `Return the Record { [[Xyz]]: _m_, [[Abc]] _xFinal_ }.`

sffc replied:
Excellent; thanks!

littledan commented on numberformat_proposed.html:
```diff
@@ -143,11 +141,15 @@ <h1>PartitionNumberPattern ( _numberFormat_, _x_ )</h1>
-        1. If _x_ is not *NaN* and _x_ < 0, then
-          1. Let _x_ be -_x_.
-          1. Let _pattern_ be _numberFormat_.[[NegativePattern]].
+        1. TODO: Deal with notation here.
```
I suppose this TODO refers to CompactDisplay and SignDisplay as well?

sffc replied:
Yes.  Thanks.  Working on these now.

littledan reviewed (commented):
This specification looks generally good to me. The organization seems clear, and I don't have any concerns about the semantics that you're exporting. I just have some editorial comments. Thanks for your patience here!

sffc committed:
Replying to second round of readme comments.

sffc committed:
Fixing up sec-intl.numberformat-internal-slots with new schema for currency and measurement units.

sffc committed:
Replying to spec review comments.  Changing references to 'tuple' to a new Record schema.

sffc committed:
Adding business logic for [[SignDisplay]].

sffc commented on numberformat_proposed.html:
```diff
@@ -202,18 +202,36 @@ <h1>PartitionNumberPattern ( _numberFormat_, _x_ )</h1>
+          1. Else if _p_ is equal `"measureUnitSymbol"` and _numberFormat_.[[UnitType]] is `"unit"`, then
+            1. Let _measureUnit_ be _numberFormat_.[[MeasureUnit]].
+            1. Let _mu_ be an ILD string representing _measureUnit_ in short form, which may depend on _x_ in languages having different plural forms. If the implementation does not have such a representation of _currency_, use _measureUnit_ itself.
+            1. Append a new Record { [[Type]]: `"unit"`, [[Value]]: _mu_ } as the last element of _result_.
```
@littledan Is there somewhere in the spec where I need to list my new [[Type]]?

littledan replied:
Nope!

sffc committed:
Adding support for [[Notation]].

sffc committed:
Adding reference locales-currencies-tz.html

sffc committed:
Adding definition for IsWellFormedUnitIdentifier

sffc commented on locales-currencies-tz.html:
```diff
@@ -0,0 +1,189 @@
+    </p>
+
+    <p>
+      All well-formed unit identifiers are allowed. However, the set of combinations of unit identifier and language tag for which localized names and symbols are available is implementation dependent. Where a localized name or symbol is not available, the unit identifier itself is used for formatting.
```
@littledan Is this OK? You can pass garbage in as your unit identifier as long as it conforms to these rules, like "hello-world", and you get output like "100 hello-world".

Is the fallback reasonable?  If a unit identifier is real but the implementation doesn't have data for it, then just use the unit identifier; for example, "100 length-meter".

littledan replied:
Thanks for calling this out; I missed it on the earlier review. I'm not sure if this is really the behavior we want. It might be better if the constructor actively signals an error if data is not available, or if the missing data is somehow visible through resolvedOptions(). Maybe we should discuss this in the Intl call. I expected IsWellFormedUnitIdentifier to check whether the unit was in the fixed set that CLDR supports.

sffc replied:
Following up in #11

littledan commented on numberformat_proposed.html:
```diff
@@ -556,7 +732,7 @@ <h1>Internal slots</h1>
       <ul>
         <li>The list that is the value of the `"nu"` field of any locale field of [[LocaleData]] must not include the values `"native"`, `"traditio"`, or `"finance"`.</li>
-        <li>[[LocaleData]][locale] must have a patterns field for all locale values. The value of this field must be a record, which must have fields with the names of the three number format styles: `"decimal"`, `"percent"`, and `"currency"`. Each of these fields in turn must be a record with the fields positivePattern and negativePattern. The value of these fields must be string values that mus ...
+        <li>[[LocaleData]][locale] must have a patterns field for all locale values. The value of this field must be a Record, which must have three fields: `"decimal"`, `"currency"`, and `"measureUnit"`. The two fields `"currency"` and `"measureUnit"` must be Records with at least one field, `"fallback"`. The `"currency"` may have additional fields with keys corresponding to currency codes accor ...
```
Editorial: This is a long paragraph. If you could break it up into inner `<ul>` tags, it'd likely be easier to read.

littledan commented on numberformat_proposed.html:
```diff
@@ -463,6 +502,143 @@ <h1>UnwrapNumberFormat( _nf_ )</h1>
+        1. Let _exponent_ be ComputeExponentForMagnitude(_numberFormat_, _magnitude_).
+        1. Let _x_ be _x_ × 10<sup>-_exponent_</sup>.
+        1. Let _formatNumberResult_ be FormatNumberToString(_numberFormat_, _x_).
+        1. If _formatNumberResult_.[[RoundedNumber]] == 0, then
```
Might this ever be -0? What is the intended behavior in that case?

`==` is discouraged in the specification; if you want to check for 0 or -0, you can use SameValueZero.

littledan commented on numberformat_proposed.html:
```diff
@@ -463,6 +502,143 @@ <h1>UnwrapNumberFormat( _nf_ )</h1>
+        1. If _x_ is 0, then
+          1. Return 0.
+        1. If _x_ &lt; 0, then
+          1. Let _x_ = -_x_.
```
s/=/be/

littledan replied:
Actually this should be, `Set _x_ to -x.`

littledan commented on numberformat_proposed.html:
```diff
@@ -721,15 +921,22 @@ <h1>Properties of Intl.NumberFormat Instances</h1>
-      <li>[[CurrencyDisplay]] is one of the String values `"code"`, `"symbol"`, or `"name"`, specifying whether to display the currency as an ISO 4217 alphabetic currency code, a localized currency symbol, or a localized currency name if formatting with the `"currency"` style. It is only used when [[Style]] has the value `"currency"`.</li>
+      <li>[[Style]] is one of the String values `"decimal"`, `"currency"`, `"percent"`, or `"unit"`. This value corresponds closely with the [[UnitType]] value.</li>
+      <li>[[UnitType]] is one of the String values `"none"`, `"currency"`, or `"measureUnit"`. The value `"measureUnit"` is used for both [[Style]] `"percent"` and `"unit"`.</li>
+      <li>[[MeasureUnit]] is a unit identifier, as defined by Unicode Technical Standard #35, Part 2, Section 6. It is only used with [[UnitType]] has the value `"measureUnit"`. The term "measure unit" is used for the purpose of the specification to distinguish between the unit type field and the manifestation of the unit type.</li>
```
If you could make the UTS 35 reference a link, it'd be even more useful.

sffc committed:
Adding link to CLDR #11271

sffc committed:
Adding reference to tc39/ecma402#244

littledan commented on locales-currencies-tz.html:
```diff
@@ -0,0 +1,189 @@
+    <h1>Measurement Unit Identifiers</h1>
+
+    <p>
+      The ECMAScript 2019 Internationalization API Specification identifies measurement units using a unit identifier as defined by Unicode Technical Standard #35, Part 2, Section 6. Their canonical form is a string containing all lowercase letters with at least one hyphen.
```
Link?

littledan commented on locales-currencies-tz.html:
```diff
@@ -0,0 +1,189 @@
+  </emu-clause>
+
+  <emu-clause id="sec-measurement-unit-identifiers">
+    <h1>Measurement Unit Identifiers</h1>
```
I guess my previous arguments that we should use "unit" rather than "measure unit" internally were unconvincing (and ditto the style/unit redundancy). This is just an editorial issue, so not a big deal one way or the other, but would you be OK with discussing this in the Intl call, to get the broader group's preferences?

sffc committed:
Rename currencySignDisplay to currencySign

sffc committed:
Adding link to TC39 Slides

sffc committed:
Fix handling of -0 in spec.

See #13. Merges change from tc39/ecma402#232 into Unified Intl.NumberFormat proposal.

sffc committed:
Generalizing 0 check to encompass -0

Logarithm of both 0 and -0 is undefined. See #13

sffc committed:
Changing == to =

Spec text uses `=` to denote equality.  #13

littledan commented:
Looks like https://github.com/sffc/proposal-unified-intl-numberformat/pull/8/files/e1f913efab1bc9797780671e0648551e5ff8b43d..7ad6fa88b64eed941e2812c589ffe36802872f4 is good behavior for -0, +1

Ms2ger committed:
Fix some typos.

Ms2ger committed:
Fix mistakes in PartitionNumberPattern().

sffc committed:
Adding rendered version of numberformat_proposed.html

sffc committed:
Merge pull request #14 from Ms2ger/typos

Fix some typos.

sffc committed:
Merge pull request #15 from Ms2ger/PartitionNumberPattern

Fix mistakes in PartitionNumberPattern().

sffc committed:
Re-generating out.html given latest PRs

sffc committed:
Moving minimumIntegerDigits resolution to FormatNumberToString, allowing for use in significant digits formatter.  Fixes #2

sffc committed:
Minor changes related to #2.

sffc committed:
Changing name of currencySignDisplay to currencySign.  Fixes #6

sffc committed:
Adding special logic for compact notation rounding.  Fixes #9

sffc committed:
Changing field names in accordance with #10.  Closes #10

sffc committed:
Adding explicit list of unit identifiers to spec.  Fixes #11

sffc committed:
Migrating build process to npm

sffc commented:
@littledan @Ms2ger @rxaviers I think I brought the spec proposal up to date with the developments from the last few months.  Could you take a look?  I am also adding it to the agenda:

https://github.com/tc39/agendas/pull/463

What else needs to happen for full Stage 3 readiness?

littledan commented:
@sffc In addition to rendering the spec, we usually produce a version with insertions and deletions explicitly noted with `<ins>` and `<del>` tags. Do you have that somewhere. This repo should also be transferred tc39 org (I sent you an email about this), and a slide deck will help in committee. You also need our reviews! It's often helpful to email your reviewers individually (and not just @-mention them), since some people forget.

sffc commented:
I am working on the `<ins><del>` version now.

sffc committed:
Adding diff version for Section 6 changes.

sffc committed:
Moving section 6 and section 11 into subdirectories.

sffc committed:
Adding Section 6 diff.  Moving files around.  Fixing index.html links.

sffc commented:
Okay!  `<ins><del>` version is ready.

https://tc39-transfer.github.io/proposal-unified-intl-numberformat/

> This repo should also be transferred tc39 org (I sent you an email about this),

I'll do that shortly.

> and a slide deck will help in committee.

I'll make one soon.

> You also need our reviews! It's often helpful to email your reviewers individually (and not just @-mention them), since some people forget.

Can you elaborate?  Is this pull request not the review?

littledan commented:
This PR is a great way to do the review, but you should probably ping people by email to make sure that they do it.

sffc committed:
Updating sffc -> tc39-transfer links in README.md

littledan commented on section11/numberformat_proposed.html:
```diff
@@ -8,28 +8,38 @@ <h1>Abstract Operations For NumberFormat Objects</h1>
       <p>
-        The abstract operation SetNumberFormatDigitOptions applies digit
-        options used for number formatting onto the intl object.
+        The abstract operation SetNumberFormatDigitOptions resolves the user-specified options in the digits realm onto the digits options object.
```
What do you mean by the term "realm" here? That's a specialized word in TC39, which I don't understand the use of here.

sffc replied:
I'm not sure where that word came from here.  This description does not need to be changed from the original.

littledan commented on section11/numberformat_proposed.html:
```diff
@@ -8,28 +8,38 @@ <h1>Abstract Operations For NumberFormat Objects</h1>
+          1. Let _mxfd_ be ? DefaultNumberOption(_mxfd_, _mnfd_, 20, _mxfdActualDefault_).
+          1. Set _intlObj_.[[MinimumFractionDigits]] to _mnfd_.
+          1. Set _intlObj_.[[MaximumFractionDigits]] to _mxfd_.
+        1. Else If _intlObj_.[[Notation]] = `"compact"`, then
```
s/=/ is /

littledan commented on section11/numberformat_proposed.html:
```diff
@@ -556,7 +752,7 @@ <h1>Internal slots</h1>
       <ul>
         <li>The list that is the value of the `"nu"` field of any locale field of [[LocaleData]] must not include the values `"native"`, `"traditio"`, or `"finance"`.</li>
-        <li>[[LocaleData]][locale] must have a patterns field for all locale values. The value of this field must be a record, which must have fields with the names of the three number format styles: `"decimal"`, `"percent"`, and `"currency"`. Each of these fields in turn must be a record with the fields positivePattern and negativePattern. The value of these fields must be string values that mus ...
+        <li>[[LocaleData]][locale] must have a patterns field for all locale values. The value of this field must be a Record, which must have three fields: `"decimal"`, `"currency"`, and `"measureUnit"`. The two fields `"currency"` and `"measureUnit"` must be Records with at least one field, `"fallback"`. The `"currency"` may have additional fields with keys corresponding to currency codes accor ...
```
This paragraph is really long! Can you break it up into an unordered sub-list for better readability?

sffc replied:
It was already really long, and I made it longer 😃.  Sure, I can break it up.  I'll do that tonight.

sffc replied:
I split it into four paragraphs.

littledan commented on section11/numberformat_proposed.html:
```diff
@@ -202,18 +225,52 @@ <h1>PartitionNumberPattern ( _numberFormat_, _x_ )</h1>
+            1. Let _cd_ be an ILD string representing _currency_ in long form, which may depend on _x_ in languages having different plural forms. If the implementation does not have such a representation of _currency_, use _currency_ itself.
             1. Append a new Record { [[Type]]: `"currency"`, [[Value]]: _cd_ } as the last element of _result_.
           1. Else,
             1. Let _literal_ be the substring of _pattern_ from position _beginIndex_, inclusive, to position _endIndex_, inclusive.
```
We should be using "unknown", not "literal", as the type here, re https://github.com/tc39/ecma402/pull/245 (though "literal" is correct for line 282).

sffc replied:
My snapshot of numberformat.html was probably taken before that PR.  I pulled in upstream and integrated any relevant changes into my spec draft.

littledan commented on section11/numberformat_proposed.html:
```diff
@@ -202,18 +225,52 @@ <h1>PartitionNumberPattern ( _numberFormat_, _x_ )</h1>
             1. Let _percentSignSymbol_ be the ILND String representing the percent sign.
             1. Append a new Record { [[Type]]: `"percentSign"`, [[Value]]: _percentSignSymbol_ } as the last element of _result_.
-          1. Else if _p_ is equal `"currency"` and _numberFormat_.[[Style]] is `"currency"`, then
+          1. Else if _p_ is equal `"measureUnitSymbol"` and _numberFormat_.[[UnitType]] is `"measureUnit"`, then
```
Why are we checking for the `[[UnitType]]` here and below? I'd suggest, instead, to add assertions inside the conditional that the `[[UnitType]]` is `measureUnit` if `_p_ equals "measureUnitSymbol"` (and you can note these assertions when you're defining the locale data).

littledan commented on section11/numberformat_proposed.html:
```diff
@@ -143,11 +162,19 @@ <h1>PartitionNumberPattern ( _numberFormat_, _x_ )</h1>
-          1. Let _pattern_ be _numberFormat_.[[NegativePattern]].
+        1. Let _exponent_ be 0.
+        1. If _x_ is *NaN*, then
+          1. Let _n_ be an ILD String value indicating the *NaN* value.
```
I don't have a suggestion for changes, but I'm wondering, do you have a good sense of when this spec uses ILD/ILND strings, and when it uses data explicitly in the locale object?

sffc replied:
My sense for ILD versus data is that ILD tends to be a small and specific string, whereas data can have many more dependencies.  For example, you could make a specific list of the ILD strings that an implementation needs to provide, but for data, the the branching factor is much larger and you can't make a specific list.

These two strings, NaN and Infinity, were already in the spec; I just moved them up a bit.

littledan replied:
BTW, why was this line moved up to this location, from its previous lower spot?

sffc replied:
I moved the calculation for _n_ and _x_ in the normal case (non-NaN-and-Infinity) up here because the new variable _exponent_ is referenced in the outer scope.  Then it looked strange that the calculation of _n_ in the special cases (NaN and Infinity) was in a different logical place than the calculation of _n_ in the normal case, so I moved those two lines up here.

littledan commented on section11/numberformat_proposed.html:
```diff
@@ -463,6 +522,143 @@ <h1>UnwrapNumberFormat( _nf_ )</h1>
+          1. Return 0.
+        1. If _x_ &lt; 0, then
+          1. Let _x_ = -_x_.
+        1. Let _magnitude_ be the greatest integer that is not greater than the base 10 logarithm of _x_.
```
I think we've decided in another thread that this should be something like, `Let _magnitude_ be the base 10 logarithm of _x_, rounded down to the nearest integer.`

sffc replied:
Will resolve when I commit against #20

littledan commented on section11/numberformat_proposed.html:
```diff
@@ -143,11 +162,19 @@ <h1>PartitionNumberPattern ( _numberFormat_, _x_ )</h1>
+          1. Let _formatNumberResult_ be FormatNumberToString(_numberFormat_, _x_).
+          1. Let _n_ be _formatNumberResult_.[[FormattedString]].
+          1. Let _x_ be _formatNumberResult_.[[RoundedNumber]].
+        1. Let _pattern_ be GetNumberFormatPattern(_numberFormat_, _x_, _exponent_)
```
It's a bit surprising that you do this calculation of which pattern to use on each format. It seems like the existing formatters choose the pattern up-front, mostly. Do you think there's anything of this that would make sense to (spec-wise) "pre-compute"? This might be useful for implementations to follow a similar split.

sffc replied:
The pattern used to be computed in the constructor.  I moved it down here.  Here are the reasons why I made this change:

1. It keeps the information stored in _numberFormat_ locale-agnostic, making it possible to do all locale data access within the format method of the spec.  Otherwise, you have some locale data access in the constructor and other locale data access in the format method, making it harder to reason about the behavior.  With this change, the constructor is in charge of "parameter resolution" only, and the format method is in charge of "locale logic" only.
2. It can be possible for the CLDR pattern itself to depend on the number and the exponent, beyond just positive, negative, and zero.  That information is not available in the constructor.  However, there are not currently locales where this happens.
3. Implementations are welcome to optimize as long as the behavior is equivalent to the spec behavior.  I wrote the spec with the intention to be clear rather than efficient.

If you feel strongly the other way, I can change it back to precompute the pattern in the constructor.

sffc replied:
Oh.  I remembered, here's the key use case.  In compact notation, you use a certain pattern for large numbers, but you fall back to the normal pattern for small numbers.  For example, in large numbers, you might be using `{number} {compactSymbol}`, but in small numbers you are just using `{number}`.

littledan commented on section11/numberformat_proposed.html:
```diff
@@ -123,14 +125,31 @@ <h1>Number Format Functions</h1>
+          1. Let _forwardZeros_ be the String consisting of _minInteger_–_intCount_ occurrences of the character `"0"`.
+          1. Let _oldString_ be _result_.[[FormattedString]].
+          1. Let _result_.[[FormattedString]] be the concatenation of Strings _forwardZeros_ and _oldString_.
+        1. Delete _result_.[[IntegerDigitsCount]].
```
We don't tend to use this kind of operation on records. Is there a way we could avoid introducing deletion to records? e.g., rather than doing so much imperative manipulation on this record, we could return a new one with just the two fields?

sffc replied:
Sounds reasonable.  Done.

littledan commented on section11/numberformat_proposed.html:
```diff
@@ -463,6 +522,143 @@ <h1>UnwrapNumberFormat( _nf_ )</h1>
+        1. Let _currencyDisplay_ be ? GetOption(_options_, `"currencyDisplay"`, `"string"`, &laquo; `"code"`, `"symbol"`, `"narrowSymbol"`, `"name"` &raquo;, `"symbol"`).
+        1. Let _unit_ be ? GetOption(_options_, `"unit"`, `"string"`, *undefined*, *undefined*).
+        1. Let _unitDisplay_ be ? GetOption(_options_, `"unitDisplay"`, `"string"`, &laquo; `"short"`, `"narrow"`, `"long"` &raquo;, `"short"`).
+        1. Let _currencySign_ be ? GetOption(_options_, `"currencySign"`, `"string"`, &laquo; `"standard"`, `"accounting"` &raquo;, `"standard"`).
```
Nit: Should this be two lines higher, for logical grouping?

sffc replied:
Done.

littledan commented on section11/numberformat_proposed.html:
```diff
@@ -143,11 +162,19 @@ <h1>PartitionNumberPattern ( _numberFormat_, _x_ )</h1>
+          1. Let _x_ be _x_ × 10<sup>-_exponent_</sup>.
+          1. Let _formatNumberResult_ be FormatNumberToString(_numberFormat_, _x_).
+          1. Let _n_ be _formatNumberResult_.[[FormattedString]].
+          1. Let _x_ be _formatNumberResult_.[[RoundedNumber]].
```
The data flow here is pretty confusing, with _x_ meaning three different things in the same straight line of code. Any way you could use different names?

sffc replied:
It could mean four different things actually:

- The raw input number
- That number scaled by 100 given a percent
- That number scaled by a power of 10 given a compact or scientific notation
- That number rounded given rounding rules

I see it as just reassigning a computer variable as you make changes to it.

The invariant that is maintained is that once you get to line 178 (after the 3-branch if statement), _x_ represents the digits displayed in the locale string.

littledan commented:
I've made some comments (and might come back with a few more), but all the issues I have are more or less editorial nit-picks. The spec text here seems to accurately capture the points that we discussed in the Intl calls and the summaries at TC39. So, I'd say this is more or less ready for Stage 3 (at which point we can continue editorial nit-picking, but it's of course nice if all issues are resolved a week and a half before the meeting).

littledan commented:
The other Stage 3 reviewers (I don't quite count since I'm the ECMA-402 editor) are @gsathya and @zenparsing.

sffc committed:
Bringing numberformat_current.html in sync with upstream.

sffc committed:
Replying to littledan feedback.

sffc committed:
Re-generating output html files.

sffc committed:
Changing base 10 logarithm language as discussed.  Fixes #20

sffc committed:
Fixing -0 condition.  Closes #13

zenparsing commented on section6/locales-currencies-tz_diff.html:
```diff
@@ -0,0 +1,332 @@
+        1. If _unitIdentifier_ begins or ends with `"-"`, return *false*.
+        1. If _unitIdentifier_ contains zero instances of `"-"`, return *false*.
+        1. If _unitIdentifier_ contains two `"-"` characters directly adjacent to one another, return *false*.
+        1. If _unitIdentifier_ is not in the following list, return *false*.
```
Does this step not imply the ones above?

sffc replied:
See #21.

zenparsing commented on section11/numberformat_diff.html:
```diff
@@ -0,0 +1,1047 @@
+      <li>[[NumberingSystem]] is a String value with the "type" given in Unicode Technical Standard 35 for the numbering system used for formatting.</li>
+      <li>[[Style]] is one of the String values `"decimal"`, `"currency"`, <del>or</del> `"percent"`<ins>, or `"unit"`</ins><del>, identifying the number format style used</del>. <ins>This value corresponds closely with the [[UnitType]] value.</ins></li>
+      <li><ins class="block">[[UnitType]] is one of the String values `"none"`, `"currency"`, or `"measureUnit"`. The value `"measureUnit"` is used for both [[Style]] `"percent"` and `"unit"`.</ins></li>
+      <li><ins class="block">[
```
`with [[UnitType]] has` => `when [[UnitType]] has`

sffc replied:
Done

zenparsing commented on section11/numberformat_diff.html:
```diff
@@ -0,0 +1,1047 @@
+      <li>[[NumberingSystem]] is a String value with the "type" given in Unicode Technical Standard 35 for the numbering system used for formatting.</li>
+      <li>[[Style]] is one of the String values `"decimal"`, `"currency"`, <del>or</del> `"percent"`<ins>, or `"unit"`</ins><del>, identifying the number format style used</del>. <ins>This value corresponds closely with the [[UnitType]] value.</ins></li>
+      <li><ins class="block">[[UnitType]] is one of the String values `"none"`, `"currency"`, or `"measureUnit"`. The value `"measureUnit"` is used for both [[Style]] `"percent"` and `"unit"`.</ins></li>
+      <li><ins class="block">[
```
`with [[UnitType]] has` => `when [[UnitType]] has`

sffc replied:
Done

zenparsing commented on section11/numberformat_diff.html:
```diff
@@ -0,0 +1,1047 @@
+      <li>[[NumberingSystem]] is a String value with the "type" given in Unicode Technical Standard 35 for the numbering system used for formatting.</li>
+      <li>[[Style]] is one of the String values `"decimal"`, `"currency"`, <del>or</del> `"percent"`<ins>, or `"unit"`</ins><del>, identifying the number format style used</del>. <ins>This value corresponds closely with the [[UnitType]] value.</ins></li>
+      <li><ins class="block">[[UnitType]] is one of the String values `"none"`, `"currency"`, or `"measureUnit"`. The value `"measureUnit"` is used for both [[Style]] `"percent"` and `"unit"`.</ins></li>
+      <li><ins class="block">[
```
Here and the line above.

```
These properties are present if and only if...
```

The internal slots are always present; maybe we want the "only used when..." phrasing?

sffc replied:
Done.

zenparsing commented on section11/numberformat_diff.html:
```diff
@@ -0,0 +1,1047 @@
+      <li><ins class="block">[[DataLocale]] is a String value with the language tag of the nearest locale for which the implementation has data to perform the formatting operation. It will be a parent locale of [[Locale]].</ins></li>
+      <li>[[NumberingSystem]] is a String value with the "type" given in Unicode Technical Standard 35 for the numbering system used for formatting.</li>
+      <li>[[Style]] is one of the String values `"decimal"`, `"currency"`, <del>or</del> `"percent"`<ins>, or `"unit"`</ins><del>, identifying the number format style used</del>. <ins>This value corresponds closely with the [[UnitType]] value.</ins></li>
+      <li><ins class="block">[[UnitType]] is one of the String values `"none"`, `"currency"`, or `"measureUnit"`. The value `"measureUnit"` is used for both [[Style]] `"percent"` and `"unit"`.</ins></li>
```
I *think* I agree with @littledan that it might be clearer to just have `[[Style]]` and introduce an abstract operation to get the "UnitType".  If that's possible!

sffc replied:
Ok.  Two votes, and there isn't a super-strong reason to keep `[[UnitType]]`.  Fixed.

zenparsing commented on section11/numberformat_diff.html:
```diff
@@ -0,0 +1,1047 @@
+    <emu-clause id="sec-setnumberformatunitoptions" aoid="SetNumberFormatUnitOptions">
+      <h1>SetNumberFormatUnitOptions ( _intlObj_, _options_ )</h1>
+      <p>
+        The abstract operation SetNumberFormatUnitOptions resolves the user-specified options in the units realm onto the unit options object.
```
The word "realm" is linking to the definition of JS Realm. Is that intentional?

sffc replied:
Fixed.

zenparsing commented on section11/numberformat_diff.html:
```diff
@@ -0,0 +1,1047 @@
+
+      <ul>
+        <li>The list that is the value of the `"nu"` field of any locale field of [[LocaleData]] must not include the values `"native"`, `"traditio"`, or `"finance"`.</li>
+        <li>[[LocaleData]][locale] must have a patterns field for all locale values. The value of this field must be a <del>record</del><ins>Record</ins>, which must have <del>fields with the names of the three number format styles: `"decimal"`, `"percent"`, and `"currency"`. Each of these fields in turn must be a record with the fields positivePattern and negativePattern.</del> <ins>three fields ...
```
I wonder if it would help to have one or more tables here describing the fields and their subfields.

sffc replied:
I split it into four separate paragraphs to make it a little easier to parse.

zenparsing reviewed (approved):
Looks good!

sffc committed:
Splitting long [[LocaleData]] patterns paragraph into four.

sffc committed:
Updating section 6 with new "core unit identifier" syntax.

sffc committed:
Removing word "realm" from SetNumberFormatUnitOptions description.

sffc committed:
Requested changes to "Properties of Intl.NumberFormat Instances"

sffc committed:
Removing [[UnitType]] and removing "measure" from internal field names.

sffc committed:
Re-rendering output HTML.

sffc commented:
zenparsing's feedback has been integrated.  Thanks for the review!

sffc committed:
Changing camelCase strings to kebab-case.  Fixes #22

sffc committed:
Updates to README file.

sffc committed:
Adding link to Stage 3 proposal slides.

sffc committed:
Restoring order of SetNumberFormatDigitOptions property reads. Fixes #27

FrankYFTang committed:
Change order of Table 2

sffc committed:
Merge pull request #30 from FrankYFTang/table2Order

Change order of Table 2

sffc committed:
Changing order in 'Properties of Intl.NumberFormat Instances' to be consistent with new ResolvedOptions order. See #28

FrankYFTang committed:
Fix link

FrankYFTang committed:
Address #31, #32, #33, #36

1. Add 5 missing units.
2. Move SetNumberFormatUnitOptions after set notation
3. Only set compactDisplay if notation is "compact"
4. Correct "concentr-percent" to "percent" in SetNumberFormatUnitOptions

sffc committed:
Merge pull request #35 from FrankYFTang/patch-1

Fix link

sffc committed:
Merge pull request #37 from FrankYFTang/order

Address #31, #32, #33, #36

sffc committed:
Adding -0 to table in README. Fixes #23

sffc committed:
Allow only simple units to be numerators and denominators. Fixes #24.

sffc committed:
Adding table for accounting format. Fixes #34

sffc committed:
Fixing another instance of core unit identifier.

sffc committed:
Updating spec to use subset of units described in #39.

sffc committed:
Update README.md

People seem to misunderstand the impact of this proposal.  I moved the boring-sounding "spec cleanup" to the bottom of the README in order to promote the measurement units and compact notation, which are the more interesting part.

Can you also update the one-line description of this repo to say the following?

> Adds localized formatting of measurement units, compact decimals, and more number features

sffc committed:
Merge pull request #42 from tc39/units-list

Updating spec to use subset of units described in #39.

sffc committed:
Merge pull request #43 from tc39/sffc-patch-1

Update README.md

sffc committed:
Adding volume units and vehicle mileage context.

sffc committed:
Change example unit to a supported unit

sffc committed:
Change unit example to common speed quantity.

sffc committed:
Add currency sign to accounting example

sffc committed:
Merge pull request #52 from tc39/sffc-patch-2

Add currency sign to accounting example

sffc committed:
Updating unit identifier syntax in README example.

sffc committed:
Removing consumption units in Section 6.

sffc committed:
Merge pull request #51 from tc39/sffc-patch-1

Change unit example to common speed quantity.

sffc committed:
Fix Record casing.

sffc committed:
Change all kebab case to camel case in numberformat Section 11.

sffc committed:
Update unit.xml link.

Fixes #49

sffc committed:
Fix casing in README.

sffc committed:
Merge pull request #54 from tc39/camelcase

Change all kebab case to camel case in numberformat Section 11.

sffc committed:
Removing spec text that entangles style: "unit" with style: "percent".

sffc committed:
Merge pull request #48 from tc39/volume

Adding volume units.

sffc committed:
Changing list to table for IsSanctionedSimpleUnitIdentifier.

sffc committed:
Merge pull request #55 from tc39/sffc-patch-1

Update unit.xml link.

sffc committed:
Compact rounding for small numbers should retain 2 significant digits.

sffc committed:
Merge pull request #59 from tc39/compact

Compact rounding for small numbers should retain 2 significant digits.

sffc committed:
Merge pull request #57 from tc39/percent

Removing spec text that entangles style: "unit" with style: "percent".

sffc committed:
Merge pull request #58 from tc39/cleanup

Change unit list to a Table, and fix casing of Record

leobalter commented:
Thanks for the updates here. It seems like in good shape as well for now. TBH, sounds easier for me to review everything from a single branch rather than as in any patch.

sffc commented:
This is the original PR that was used for Stage 3 review.  Any further changes from Stage 3 have been separate PRs.  I'll close this PR to reduce confusion.
